Refactor shared code

adrianna-chang-shopify opened the pull request:
Extracts various pieces of shared code into modules. Rather than extract based on the object that implements it (ie. "Module-code", "Class-code"), we extract into modules based on a grouping of behaviour (ie. constants, methods, instance variables, etc.)

I've left the singleton methods untouched for now, as I suspect those will be more difficult to deal with.

tomstuart reviewed (approved):
Yeah, this is great! A couple of nitpicks but nothing big. :shipit: 

So what next in our quest for a nice rigorous implementation of the functionality we’ve seen thus far? Superficially, `#kind_of?` is an obvious remaining piece of duplicated code which we may wish to clean up. Perhaps more significantly, we’re now in a better position to be able to fill any accidental gaps which exist in our implementation (and aren’t checked by the tests); for example, none of these currently work:

* `Toy::Object.instance_variables`, `Toy::Module.instance_variables`, `Toy::Class.instance_variables` (et al)
* `Toy::Object.constants`, `Toy::Module.constants`, `Toy::Class.constants` (et al)
* `Toy::Object.instance_methods`, `Toy::Module.instance_methods`, `Toy::Class.instance_methods` (et al)
* `Toy::Object.new.to_s`, `Toy::Module.new.to_s`, `Toy::Class.new.to_s`
* `Toy::Object.new.inspect`
* `Toy::Object.new.kind_of?`, `Toy::Module.new.kind_of?`, `Toy::Class.new.kind_of?`

Now that we have mixins which provide (some of) these, can we plug all the holes?

Once we’ve plugged all the holes, we can finish up by figuring out the best way to organise all this shared code so that the right functionality ends up in the right places in the clearest way possible.

(As ever, I greatly appreciate your patience! 😅)

adrianna-chang-shopify commented on lib/toy/behaviours.rb:
```diff
@@ -25,5 +25,36 @@ def remove_instance_variable(name)
     end
+
+    module Constants
+      def inspect
+        "#<#{self.class}>"
+      end
```
Yep, totally fair! I was considering it a requirement of those constant methods rather than a part of the Object API, but it makes more sense to treat this as part of the Ruby API that we're trying to implement in our Toy object space 👍

adrianna-chang-shopify commented on lib/toy/class.rb:
```diff
@@ -14,30 +15,7 @@ def Class.new(superclass = Object)
-      def remove_instance_variable(name)
-        ivar_map.delete(name.to_sym)
-      end
+      include Behaviours::InstanceVariables
```
I can clean this up in a rebase 😄

adrianna-chang-shopify commented:
As always, thanks for the feedback Tom! ❤️  I've addressed the comments in this PR -- I'll tackle plugging the remaining holes in a follow up PR, since I'm considering this PR to be more of a pure refactor. I'll open an issue linking to your comment 😄

tomstuart commented on lib/toy/class.rb:
```diff
@@ -96,29 +74,7 @@ def new
-          def remove_instance_variable(name)
-            ivar_map.delete(name.to_sym)
-          end
+          include Behaviours::InstanceVariables
```
Thanks for the rebase! 😊
